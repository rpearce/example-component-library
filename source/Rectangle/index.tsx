import React, { FC } from 'react'

interface Props {
  desc?: string
  fill?: string
  height: number
  title?: string
  width: number
}

const Rectangle: FC<Props> = ({ desc, fill, height, title, width }) => (
  <svg
    className="rl-rect"
    fill={fill}
    height={height}
    role="img"
    viewBox={`0 0 ${width} ${height}`}
    width={width}
    xmlns="http://www.w3.org/2000/svg"
  >
    {title && <title>{title}</title>}
    {desc && <desc>{desc}</desc>}
    <rect width={width} height={height} />
  </svg>
)

Rectangle.defaultProps = {
  height: 100,
  width: 100,
}

export default Rectangle
