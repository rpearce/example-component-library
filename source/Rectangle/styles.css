.rl-rect { fill: #30336b; }
