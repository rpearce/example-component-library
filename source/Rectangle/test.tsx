import React from 'react'
import { render } from '@testing-library/react'
import axe from 'axe-core'
import Rectangle from './index'

test('with all props', () => {
  const { asFragment, container, getByText } = render(
    <Rectangle
      className="class-override"
      desc="A rectangle that is 4 times taller than it is wide"
      height={600}
      title="Minimalist building"
      width={150}
    />
  )

  const svgEl = container.querySelector('svg')
  const titleEl = getByText('Minimalist building')
  const descEl = getByText('A rectangle that is 4 times taller than it is wide')

  expect(svgEl).toHaveAttribute('height', '600')
  expect(svgEl).toHaveAttribute('width', '150')
  expect(titleEl).toBeInTheDocument()
  expect(descEl).toBeInTheDocument()
  expect(asFragment()).toMatchSnapshot()
})

test('with only height, title, width', () => {
  const { asFragment, container, getByText } = render(
    <Rectangle height={600} title="Minimalist building" width={150} />
  )
  const svgEl = container.querySelector('svg')
  const titleEl = getByText('Minimalist building')
  const descEl = container.querySelector('desc')

  expect(svgEl).toHaveAttribute('height', '600')
  expect(svgEl).toHaveAttribute('width', '150')
  expect(titleEl).toBeInTheDocument()
  expect(descEl).not.toBeInTheDocument()
  expect(asFragment()).toMatchSnapshot()
})

test('is accessible with height, width, title, desc', (done) => {
  const { container } = render(
    <Rectangle
      desc="A rectangle that is 4 times taller than it is wide"
      height={600}
      title="Minimalist building"
      width={150}
    />
  )

  axe.run(container, {}, (err, result) => {
    expect(err).toEqual(null)
    expect(result.violations.length).toEqual(0)
    done()
  })
})

test('is inaccessible without title', (done) => {
  const { container } = render(
    <Rectangle
      desc="A rectangle that is 4 times taller than it is wide"
      height={600}
      title="Minimalist building"
      width={150}
    />
  )

  // do something very wrong to prove a11y testing
  container.querySelector('title')?.remove()

  axe.run(container, {}, (err, result) => {
    expect(err).toEqual(null)
    expect(result.violations[0].id).toEqual('svg-img-alt')
    done()
  })
})
