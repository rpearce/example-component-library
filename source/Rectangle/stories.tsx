import React from 'react'
import { storiesOf } from '@storybook/react'
import { withA11y } from '@storybook/addon-a11y'
import { color, number, text, withKnobs } from '@storybook/addon-knobs'
import '../../.storybook/base.css'
import './styles.css'
import { Rectangle } from '../../dist/esm'

const stories = storiesOf('Rectangle', module)

stories.addDecorator(withA11y)
stories.addDecorator(withKnobs)

stories.add('default', () => <Rectangle height={100} width={400} />)
