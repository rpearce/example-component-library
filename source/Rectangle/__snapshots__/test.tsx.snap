// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with all props 1`] = `
<DocumentFragment>
  <svg
    class="rl-rect"
    height="600"
    role="img"
    viewBox="0 0 150 600"
    width="150"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      Minimalist building
    </title>
    <desc>
      A rectangle that is 4 times taller than it is wide
    </desc>
    <rect
      height="600"
      width="150"
    />
  </svg>
</DocumentFragment>
`;

exports[`without title & desc 1`] = `
<DocumentFragment>
  <svg
    class="rl-rect"
    height="600"
    role="img"
    viewBox="0 0 150 600"
    width="150"
    xmlns="http://www.w3.org/2000/svg"
  >
    <rect
      height="600"
      width="150"
    />
  </svg>
</DocumentFragment>
`;
